### PR TITLE
Post Interview Feedback

### PR DESCRIPTION
**Pull request by NicholasMatias**
Features:
- Now a text classifier model for situation, task, action, and result => each model returns a score from 0 to 1.
- Each model is trained off of a dataset that contains 10,000 user responses to a wide array of behavioral interview questions => responses are broken up into situation, task, action, and result.
- These models are used to evaluate the user's response to a given question => on the frontend, STAR scores are color-coded. => color is decided by score. 
- Relevance check => algorithm takes in the question and the user response => uses cosine similarity to determine if the user response is relevant to the given question. 
- Most frequently used words => returns the 10 most frequently used words along with their frequencies
- Most frequently used phrases => returns the 5 most used phrases (two word pairings) => may change the gramSize to 3 or simply add another call that uses a gramSize of 3. => gramSize determines how many words are looked at in sequence for frequency
- Tooltips have been added in to explain what the STAR method is. => gives context to why InterviewMe is evaluating user responses based off of the STAR method. 
- Tooltip for star, situation, task, action, result, and relevancy. => explains what they mean in the context of the scoring. 
- Loading state has been added for the interview feedback. 
- In prep for questions page => added to nav bars on all pages => page initialized (blank page). 

Interview. Feedback Diagram: [https://internalfb.com/excalidraw/EX253465](https://internalfb.com/excalidraw/EX253465)

SS 1:
Screenshot 2024-07-23 at 5 02 04 PM
SS 2:
Screenshot 2024-07-23 at 5 03 10 PM
SS 3:
Screenshot 2024-07-23 at 5 12 43 PM

SS 4:
Screenshot 2024-07-23 at 5 10 40 PM

### CODE REVIEW

File: backend/models/TextClassification.cjs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+	const situation_prediction = situation_model.predict(userResponse)
+	const task_prediction = task_model.predict(userResponse)
+	const action_prediction = action_model.predict(userResponse)
+	const result_prediction = result_model.predict(userResponse)
+
+	
+	return [situation_prediction[0].confidence, task_prediction[0].confidence, action_prediction[0].confidence, result_prediction[0].confidence]
+}
+
+function getSituation(userResponse) {
```
Comment (ocorrodi) [inline backend/models/TextClassification.cjs:37]: these functions all seem pretty similar- can we consolidate into one generic function and pass in situation/action/task/result and the corresponding model (or get that from which part of star is passed in)?

File: backend/models/QnA_Relevance.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+    const questionVector = tf_idf.listTerms(0).map(term => term.tfidf)
+    const responseVector = tf_idf.listTerms(1).map(term => term.tfidf)
+
+    while (questionVector.length < responseVector.length) questionVector.push(0);
+    while (responseVector.length < questionVector.length) responseVector.push(0);
+
+    const similarity = cosineSimilarity(questionVector, responseVector)
+
+    if (!similarity || similarity <= .3) {
```
Comment (ocorrodi) [inline backend/models/QnA_Relevance.js:48]: let's move the thresholds to constants?

File: backend/models/QnA_Relevance.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    if (magA === 0 || magB === 0) {
+        return 0;
+    }
+
+    const cosSim = dotProd / (magA * magB);
+    return cosSim;
+}
+
+export function getRelevanceScore(question, response){
```
Comment (ocorrodi) [inline backend/models/QnA_Relevance.js:34]: would be great to have a link to diagram or comment explaining logic for the score/how it's calculated!

**Review by ocorrodi — COMMENTED** [review]
nice work! cool to see the synthesis from the qualitative concepts about what makes a good response into a quantitative implementation/score.

File: backend/models/Word_Freq.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import natural from 'natural'
+const { WordTokenizer, PorterStemmer } = natural
+
+export function cleanAndTokenize(text) {
+    const stopWords = new Set([" ","m","i", "me", "my", "myself", "we", "our", "ours", "ourselves", "you", "your", "yours", "yourself", "yourselves", "he", "him", "his", "himself", "she", "her", "hers", "herself", "it", "its", "itself", "they", "them", "their", "theirs", "themselves", "what", "which", "who", "whom", "this", "that", "these", "those", "am", "is", "are", "was", "were", "be", "been", "being", "have", "has", "had", "having", "do", "does", "did", "doing", "a", "an", "the", "and", "but", "if", "or", "because", "as", "until", "while", "of", "at", "by", "for", "with", "about", "against", "between", "into", "through", "during", "before", "after", "above", "below", "to", "from", "up", "down", "in", "out", "on", "off", "over", "under", "again", "further", "then", "once", "here", "there", "when", "where", "why", "how", "all", "any", "both", "each", "few", "more", "most", "other", "some", "such", "no", "nor", "not", "only", "own", "same", "so", "than", "too", "very", "s", "t", "can", "will", "just", "don", "should", "now"])
```
Comment (jgsiqueira) [inline backend/models/Word_Freq.js:5]: nit: have this in a json file or loaded from db

File: backend/models/Word_Freq.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+export function countFreq(words) {
+    const freq = {}
+    words.forEach(word => {
+        freq[word] = (freq[word] || 0) + 1
+    })
+    return freq
+}
+
+export function getFreq(text, n = 10, nGramSize = 1) {
+    const words = nGramSize===1 ? cleanAndTokenize(text) : cleanAndTokenizeNGrams(text)
```
Comment (jgsiqueira) [inline backend/models/Word_Freq.js:39]: I've seen inconsistent formatting in the last PRs.
Were you able to install an auto-formatter? (early feedback I gave in the program)

Comment (NicholasMatias) [reply backend/models/Word_Freq.js:39]: I went through all of the files to check and you are correct. The auto-formatter hasn't been running; the extension needed an update on VSCode. I have been using prettier as the auto formatter but it seems it hasn't been applying in recent files due to needing an update. That said, the issue has been resolved and prettier is now back up and running correctly. All current and future code will be formatted accordingly.

**Pull request merged**